### PR TITLE
Changes to use moq-netcore

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.Core.Test/Controllers/DefaultControllerFactoryTest.cs [hunk 1]
```diff
@@ -101,15 +102,23 @@ public void CreateController_SetsBindingContext()
                 ControllerTypeInfo = typeof(ControllerWithAttributes).GetTypeInfo()
             };
             var bindingContext = new ActionBindingContext();
-
+            var bindingContextAccessor = new ActionBindingContextAccessor
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/Controllers/DefaultControllerFactoryTest.cs]: This used to work because it was relying on `CallContext`

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CollectionModelBinderTest.cs [hunk 2]
```diff
@@ -459,7 +451,7 @@ private class ListWithThrowingConstructor<T> : List<T>
             public ListWithThrowingConstructor()
                 : base()
             {
-                throw new ApplicationException("No, don't do this.");
+                throw new RankException("No, don't do this.");
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CollectionModelBinderTest.cs]: More like `DankException`

File: test/Microsoft.AspNet.Mvc.DataAnnotations.Test/DataAnnotationsModelValidatorTest.cs [hunk 3]
```diff
@@ -64,15 +58,15 @@ public static IEnumerable<object[]> Validate_SetsMemberName_OnValidationContext_
             string expectedMemberName)
         {
             // Arrange
-            var attribute = new Mock<ValidationAttribute> { CallBase = true };
-            attribute.Protected()
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.DataAnnotations.Test/DataAnnotationsModelValidatorTest.cs]: Moq.Protected probably works correctly in CoreCLR now (the bug was addressed), but the change looks nicer + is strongly typed.

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CollectionModelBinderTest.cs (hunk 2 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CollectionModelBinderTest.cs]: #YOLO

File: test/Microsoft.AspNet.Mvc.Core.Test/Controllers/ControllerActionInvokerTest.cs [hunk 4]
```diff
@@ -31,7 +31,7 @@ namespace Microsoft.AspNet.Mvc.Controllers
     public class ControllerActionInvokerTest
     {
         // Intentionally choosing an uncommon exception type.
-        private readonly Exception _actionException = new TimeZoneNotFoundException();
+        private readonly Exception _actionException = new DivideByZeroException();
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/Controllers/ControllerActionInvokerTest.cs]: ??

File: test/Microsoft.AspNet.Mvc.Core.Test/Controllers/DefaultControllerFactoryTest.cs (hunk 1 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/Controllers/DefaultControllerFactoryTest.cs]: It used to work because `GetServices()` set up the `IActionBindingContextAccessor` instance. Why was that removed?

File: test/Microsoft.AspNet.Mvc.Core.Test/FileResultTest.cs [hunk 5]
```diff
@@ -241,7 +240,7 @@ private class EmptyFileResult : FileResult
             public bool WasWriteFileCalled;
 
             public EmptyFileResult()
-                : this(MediaTypeNames.Application.Octet)
+                : this("application/octet")
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/FileResultTest.cs]: ??

File: test/Microsoft.AspNet.Mvc.Core.Test/HttpResponseStreamWriterTest.cs [hunk 6]
```diff
@@ -467,14 +480,13 @@ private class TestMemoryStream : MemoryStream
...
             private int _flushAsyncCallCount;
-            private int _closeCallCount;
             private int _disposeCallCount;
 
             public int FlushCallCount { get { return _flushCallCount; } }
 
             public int FlushAsyncCallCount { get { return _flushAsyncCallCount; } }
 
-            public int CloseCallCount { get { return _closeCallCount; } }
+            public int CloseCallCount { get; private set; }
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/HttpResponseStreamWriterTest.cs]: ??

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CollectionModelBinderTest.cs (hunk 2 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CollectionModelBinderTest.cs]: Why `RankException` and not e.g. `AggregateException`?

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/DictionaryModelBinderTest.cs [hunk 7]
```diff
@@ -562,9 +561,8 @@ private class DictionaryWithThrowingConstructor<TKey, TValue> : Dictionary<TKey,
             public DictionaryWithThrowingConstructor()
                 : base()
             {
-                throw new ApplicationException("No, don't do this.");
+                throw new Exception("No, don't do this.");
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/DictionaryModelBinderTest.cs]: Be consistent w/ `ApplicationException` replacements.

File: test/Microsoft.AspNet.Mvc.DataAnnotations.Test/DataAnnotationsModelValidatorTest.cs [hunk 8]
```diff
@@ -64,15 +58,15 @@ public static IEnumerable<object[]> Validate_SetsMemberName_OnValidationContext_
...
-                     .Verifiable();
+            var attribute = new Mock<TestableValidationAttribute> { CallBase = true };
+            attribute
+                .Setup(p => p.IsValidPublic(It.IsAny<object>(), It.IsAny<ValidationContext>()))
+                .Callback((object o, ValidationContext context) =>
+                {
+                    Assert.Equal(expectedMemberName, context.MemberName);
+                })
+                .Returns(ValidationResult.Success)
+                .Verifiable();
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.DataAnnotations.Test/DataAnnotationsModelValidatorTest.cs]: To confirm: Core Moq doesn't support `.Protected()`?

File: test/Microsoft.AspNet.Mvc.DataAnnotations.Test/TestResources.cs [hunk 9]
```diff
@@ -14,10 +15,17 @@ public static class TestResources
 
         public static string DisplayAttribute_Name { get; } = Resources.DisplayAttribute_Name;
 
+        public static CultureInfo CurrentUICulture =>
+#if DNX451
+            Thread.CurrentThread.CurrentUICulture;
+#else
+            CultureInfo.CurrentUICulture;
+#endif
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.DataAnnotations.Test/TestResources.cs]: Why is this necessary? It's only the setters that differ.

File: test/Microsoft.AspNet.Mvc.Core.Test/Controllers/ControllerActionInvokerTest.cs (hunk 4 above)
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.Core.Test/Controllers/ControllerActionInvokerTest.cs]: Probably not available on CoreCLR.

File: test/Microsoft.AspNet.Mvc.TestCommon/TestFileChangeToken.cs [hunk 10]
```diff
@@ -5,7 +5,7 @@
 
 namespace Microsoft.Extensions.Primitives
 {
-    internal class TestFileChangeToken : IChangeToken
+    public class TestFileChangeToken : IChangeToken
```
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.TestCommon/TestFileChangeToken.cs]: This is a test project.

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TestCommon/TestFileChangeToken.cs]: This is a test _shared source_ project. Should be no need for any access beyond `internal`.

Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.TestCommon/TestFileChangeToken.cs]: Yeah I get that, but what difference does it make?

File: test/Microsoft.AspNet.Mvc.Core.Test/Controllers/DefaultControllerFactoryTest.cs (hunk 1 above)
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/Controllers/DefaultControllerFactoryTest.cs]: That doesn't get flown in to the controller factory.

File: test/Microsoft.AspNet.Mvc.Core.Test/FileResultTest.cs (hunk 5 above)
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/FileResultTest.cs]: Doesn't exist in coreclr.

File: test/Microsoft.AspNet.Mvc.Core.Test/Controllers/DefaultControllerFactoryTest.cs (hunk 1 above)
Comment (davidfowl) [inline test/Microsoft.AspNet.Mvc.Core.Test/Controllers/DefaultControllerFactoryTest.cs]: flowed* /cc @JunTaoLuo

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CollectionModelBinderTest.cs (hunk 2 above)
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CollectionModelBinderTest.cs]: Quite likely that a Task.Result can throw `AggregateException`. This ensures it's unique.

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/DictionaryModelBinderTest.cs (hunk 7 above)
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/DictionaryModelBinderTest.cs]: Ok.

File: test/Microsoft.AspNet.Mvc.DataAnnotations.Test/DataAnnotationsModelValidatorTest.cs (hunk 8 above)
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.DataAnnotations.Test/DataAnnotationsModelValidatorTest.cs]: It does - there was a bug when I initially ported it which apparently was later fixed. Haven't seen if it works since the fix

File: test/Microsoft.AspNet.Mvc.DataAnnotations.Test/TestResources.cs (hunk 9 above)
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.DataAnnotations.Test/TestResources.cs]: Don't think there's `CurrentThread.CurrentUICulture` in CoreCLR.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/project.json [hunk 11]
```diff
@@ -3,7 +3,7 @@
     "../Microsoft.AspNet.Mvc.Formatters.Xml.Test/XmlAssert.cs"
   ],
   "compilationOptions": {
-    "define": [ "__RemoveThisBitTo__GENERATE_BASELINES" ],
+    "define": [ "__RemoveThisBitTo__GENERATE_BASELINES", "FUNCTIONAL_TESTS" ],
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/project.json]: One of the types in TestCommon gets included here and it used to exclude Moq out using `#if MOQ_SUPPORT`. This replaces it.

File: test/Microsoft.AspNet.Mvc.TestCommon/TestFileChangeToken.cs (hunk 10 above)
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.TestCommon/TestFileChangeToken.cs]: Internal types + strong naming doesn't work with Moq NetCore now. I don't see us referencing one test project from another so not seeing how this would be problematic. If it is, we can change this into a regular project instead.

File: test/Microsoft.AspNet.Mvc.Formatters.Xml.Test/XmlDataContractSerializerOutputFormatterTest.cs [hunk 12]
```diff
@@ -426,6 +426,7 @@ public void GetSupportedContentTypes_ReturnsSupportedTypes(Type type, object exp
             }
         }
 
+#if !DNXCORE50
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Formatters.Xml.Test/XmlDataContractSerializerOutputFormatterTest.cs]: Sure - let me figure out why these couldn't be run.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/project.json (hunk 11 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/project.json]: Why does that class need to exclude Moq?

File: test/Microsoft.AspNet.Mvc.DataAnnotations.Test/TestResources.cs (hunk 9 above)
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.DataAnnotations.Test/TestResources.cs]: The property doesn't exist on Thread.CurrentThread.

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.DataAnnotations.Test/TestResources.cs]: So why not use `CultureInfo.CurrentUICulture` everywhere?

File: test/Microsoft.AspNet.Mvc.FunctionalTests/project.json (hunk 11 above)
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/project.json]: We don't use Moq in functional tests.

File: test/Microsoft.AspNet.Mvc.Core.Test/HttpResponseStreamWriterTest.cs (hunk 6 above)
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/HttpResponseStreamWriterTest.cs]: The field is unused in CoreCLR which results in a warning \ error.

File: src/Microsoft.AspNet.Mvc.WebApiCompatShim/HttpRequestMessage/HttpRequestMessageExtensions.cs [hunk 13]
```diff
@@ -331,7 +331,10 @@ public static HttpResponseMessage CreateResponse<T>(this HttpRequestMessage requ
             if (result?.Formatter == null)
             {
                 // Return a 406 when we're actually performing conneg and it fails to find a formatter.
-                return request.CreateResponse(HttpStatusCode.NotAcceptable);
+                return new HttpResponseMessage(HttpStatusCode.NotAcceptable)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.WebApiCompatShim/HttpRequestMessage/HttpRequestMessageExtensions.cs]: cc @rynowak - the `request.CreateResponse` doesn't exist in CoreCLR and it ended up recursively calling this method.

File: test/Microsoft.AspNet.Mvc.Razor.Test/Compilation/ExpressionRewriterTest.cs [hunk 14]
```diff
@@ -456,10 +457,20 @@ public IEnumerable<FieldDeclarationSyntax> FindFields(SyntaxNode node)
 
         private CSharpCompilation Compile(SyntaxTree tree)
         {
+            // Disable 1702 until roslyn turns this off by default
+            var options = new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary)
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Razor.Test/Compilation/ExpressionRewriterTest.cs]: Copied from https://github.com/aspnet/dnx/blob/167fff5c19a0864dcf4b31edaf7c17d13bd35a3f/src/Microsoft.Dnx.Compilation.CSharp.Common/CompilerOptionsExtensions.cs#L30-L34 since Roslyn prints assembly redirect warnings in CoreCLR.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewFeatures/ViewDataDictionaryTest.cs [hunk 15]
```diff
@@ -682,7 +681,14 @@ public void EvalWithNestedDictionariesEvalCorrectly()
         {
             // Arrange
             var viewData = new ViewDataDictionary(new EmptyModelMetadataProvider());
-            viewData.Add("Foo", new Dictionary<string, object> { { "Bar", new Hashtable { { "Baz", "Quux" } } } });
+            var value = new Dictionary<string, object>
+            {
+                ["Bar"] = new Dictionary<string, string>
+                {
+                    { "Baz", "Quux" }
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewFeatures/ViewDataDictionaryTest.cs]: Unfortunately I'm can only support half Pranav points here. Sorry.

**Comment by dougbu** [comment]
:shipit:, preferably with a couple more comment near excluded tests (comments still open above)

**Comment by Eilon** [comment]
:shipit:
